### PR TITLE
fix(tokenizer): support Kimi K2/K2.5/K2.6 tiktoken models

### LINKED ISSUE #1475: tiktoken decode panics on unmapped Kimi K2.5 reserved token IDs
## Summary

`TiktokenTokenizer::decode` in the smg gateway panics with `no entry found for key` when it has to decode a Kimi-K2.5 token id that lives in the vocab but is not registered by `load_from_path`. The panic crashes a tokio worker thread and breaks the in-flight `/v1/chat/completions` connection. Hit reliably in long sweeps (BFCL non_live, 16 panics over 1390 requests against `nvidia/Kimi-K2.5-NVFP4`).

This is a smg-side tokenizer wiring bug — the engine emits a valid id within the model's declared vocab range (`vocab_size = 163840`), and smg's decoder is the one that fails to handle it.

## Symptom

```
[smg] thread 'tokio-rt-worker' (1724695) panicked at
  /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tiktoken-rs-0.9.1/src/patched_tiktoken.rs:82:64:
  no entry found for key
```

Many lossy-UTF-8 warns precede the panic (normal multi-byte boundary handling), then one specific token id triggers the unconditional index access.

## Root cause

`tiktoken-rs-0.9.1/src/patched_tiktoken.rs` (`_decode_native_and_split`):

```rust
let token_bytes = self
    .decoder
    .get(&token)
    .unwrap_or_else(|| &self.special_tokens_decoder[&token]);  // panics if missing
```

If `token` is in neither `decoder` nor `special_tokens_decoder`, the `[&token]` index panics. The crate exposes no safe variant.

smg's lossy-decode fallback in `crates/tokenizer/src/tiktoken.rs:473-494` (`Decoder for TiktokenTokenizer::decode`) is the only caller exposed to this path.

## Kimi K2.5 vocab gap

`TiktokenTokenizer::load_from_path` (`crates/tokenizer/src/tiktoken.rs:217-251`) builds `special_tokens_encoder` from only two sources:

| Source | IDs covered |
|---|---|
| `tiktoken.model` BPE encoder | 163584 entries (ids `0..163583`) |
| `tokenizer_config.json` `added_tokens_decoder` | 23 entries scattered in `163584..163839` |

That leaves **233 ids unmapped** in `{163589, 163592, 163600, 163608..163837}`. These are reserved/extra-id slots in Kimi-K2.5's declared 163840-token vocab. They're rare in normal sampling but not impossible — over a long sweep they show up and crash a worker.

## Reproducer (Python parity of the panicking lookup)

Loads the same two sources smg loads, runs the same fallback `_decode_native_and_split` logic:

```python
import json, base64
ranks = {}
with open('/path/to/Kimi-K2.5-NVFP4/tiktoken.model') as f:
    for line in f:
        line = line.strip()
        if not line: continue
        b64, r = line.split()
        ranks[int(r)] = base64.b64decode(b64)

with open('/path/to/Kimi-K2.5-NVFP4/tokenizer_config.json') as f:
    cfg = json.load(f)
atd = {int(k): v['content'].encode('utf-8') for k, v in cfg['added_tokens_decoder'].items()}

def decode_native_and_split(token_ids):
    for t in token_ids:
        if t in ranks:
            yield ranks[t]
        else:
            yield atd[t]   # KeyError ≡ Rust panic at patched_tiktoken.rs:82:64

next(decode_native_and_split([163600]))  # KeyError: 163600
next(decode_native_and_split([163700]))  # KeyError: 163700
```

Both ids are within the model's declared `vocab_size` (163840) and are valid model output ids.

## Environment

- smg `1.4.2.dev15` (from `lightseek.org/whl/cu130/`)
- smg-grpc-servicer `0.5.3.dev15`, smg-grpc-proto `0.4.8.dev15`
- Model: `nvidia/Kimi-K2.5-NVFP4`
- Workload: BFCL non_live, 1390 prompts, `--num-threads 64`
- Frequency: 16 panics across 1390 requests; 3 surfaced to the client after BFCL/tenacity retries

## Notes

- This is independent of the backend: any backend that ever emits one of the 233 reserved ids will trip the same panic.
- The lossy-UTF-8 fallback is fine; the panic is in `_decode_native_and_split`, called from the fallback before the warn is logged.
- The panicking worker takes down only one tokio worker; the smg process keeps running, but the affected client request returns `Server disconnected without sending a response`.

### PR DESCRIPTION
**Pull request by CatherineSue**
### Problem

Kimi K2 / K2.5 / K2.6 ship tiktoken-style vocabs alongside their HF repos. Their `tokenizer_config.json` enumerates a handful of named special tokens (e.g. `<|im_assistant|>` at ID 163588), but does **not** enumerate every slot in the 256-slot reserved-special-token range that upstream `tokenization_kimi.py` synthesizes via:

```python
{special_tokens_mapping.get(i, f"<|reserved_token_{i}|>"): i
 for i in range(num_base_tokens, num_base_tokens + 256)}
```

The generic `TiktokenTokenizer` only registered IDs present in `tiktoken.model` or `added_tokens_decoder`, so unnamed reserved IDs were unknown to `CoreBPE`. Decoding a sequence that contained one fell through into `_decode_native_and_split`, which indexes `special_tokens_decoder[&token]` directly — **panic** on the missing key.

There is also a pre-tokenizer mismatch: Kimi's regex begins with `[\p{Han}]+` and uses character-class intersection (`[X&&[^\p{Han}]]`) so Han chars are split from the letter/contraction branches. The generic path used the cl100k regex, mangling Han-heavy inputs.

Fixes #1475.

### Solution

Detect Kimi inside `TiktokenTokenizer::load_from_path` and specialize two things:
1. **Pattern** — use the Han-aware regex (byte-identical to upstream `pat_str`, verified against K2-Thinking, K2.5, and K2.6 — all three `tokenization_kimi.py` files are identical on the BPE-relevant fields).
2. **Reserved slots** — fill 256 entries at `[len(mergeable_ranks), len(mergeable_ranks)+256)` with `<|reserved_token_{id}|>`, skipping any IDs already occupied by explicit `added_tokens_decoder` entries.

Detection lives in a crate-private `kimi_k2_tokenizer` helper module and exposes only three items to `tiktoken.rs`: the regex constant, `matches_dir`, and `apply_reserved_special_tokens`. There is no separate public `KimiK2Tokenizer` type and no change to `factory.rs` — Kimi support is transparent at the call surface, so future tiktoken variants (LLaMA-3-tiktoken, etc.) can follow the same pattern without growing a dispatch table.

Detection matches all three Kimi variants:
- Primary: `tokenizer_config.json` mentions `tokenization_kimi` (via `auto_map`, `tokenizer_class`, etc.).
- Fallback: `config.json::model_type` ∈ `{kimi_k2, kimi_k25}`. K2-Thinking reports `kimi_k2`; K2.5 and K2.6 both report `kimi_k25`. The architecture string is intentionally **not** checked — K2-Thinking shares `DeepseekV3ForCausalLM` with vanilla DeepSeek and would false-positive.

Separately, the generic tiktoken decode path now returns a clean `Err` on truly unknown token IDs instead of falling into the lossy-byte path that panics. The guard matches tiktoken-rs's `DecodeKeyError` Display message; the typed error lives in a private module (`vendor_tiktoken`) and isn't re-exported as of tiktoken-rs 0.9.1, so a typed downcast isn't available.

## Changes

- `crates/tokenizer/src/kimi_k2_tokenizer.rs` (new, crate-private): `KIMI_K2_PATTERN`, `matches_dir`, `apply_reserved_special_tokens`, and helpers for the two-tier detection.
- `crates/tokenizer/src/tiktoken.rs`: `load_from_path` now consults the Kimi module, picks the pattern, pre-fills reserved slots before constructing `CoreBPE`; `from_encoder_and_config` is now private. `decode` returns a structured error on unknown IDs instead of letting `_decode_native_and_split` panic.
- `crates/tokenizer/src/lib.rs`: `mod kimi_k2_tokenizer` (was `pub mod`); the helper module has no public surface.

## Test Plan

`cargo test -p llm-tokenizer --lib` — 132/132 pass, including five new Kimi tests:

- `reserved_special_tokens_are_synthesized` — 256 slots filled, explicit names preserved, vocab_size correct, decode/encode of synthetic reserved IDs round-trips.
- `matches_via_model_type_kimi_k2` — `model_type: "kimi_k2"` (K2-Thinking) triggers Kimi load path.
- `matches_via_model_type_kimi_k25` — `model_type: "kimi_k25"` (K2.5 + K2.6) triggers Kimi load path.
- `uses_kimi_pattern_for_contractions` — the Kimi regex matches `hello's` as one chunk (cl100k would split before `'s`).
- `han_input_round_trips_through_kimi_pattern` — Han + ASCII mixed input encodes without panic and decodes byte-equal. Guards the fancy-regex character-class-intersection compatibility risk.

`cargo clippy -p llm-tokenizer --lib --tests --no-deps` clean.
`cargo +nightly fmt -p llm-tokenizer -- --check` clean.

<details>
<summary>Checklist</summary>

- [x] `cargo +nightly fmt` passes
- [x] `cargo clippy --all-targets --all-features -- -D warnings` passes
- [ ] (Optional) Documentation updated
- [ ] (Optional) Please join us on Slack [#sig-smg](https://slack.lightseek.org) to discuss, review, and merge PRs

</details>

## Summary by CodeRabbit

* **New Features**
  * Automatic detection and support for Kimi K2/K2.5/K2.6 tokenizers, including Han-aware tokenization and improved contraction handling.
  * Synthesized reserved special-token ranges to avoid ID conflicts.

* **Bug Fixes**
  * Clearer user-facing error when decoding unknown token IDs.
  * More robust loading when tokenizer config file is absent.

* **Tests**
  * Added tests for tokenizer detection, reserved-token synthesis, missing-config handling, and decode error behavior.

[Review Change Stack](https://app.coderabbit.ai/change-stack/lightseekorg/smg/pull/1482)

### CODE REVIEW

File: crates/tokenizer/src/kimi_k2_tokenizer.rs [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+fn read_json(path: &Path) -> Option<Value> {
+    let content = std::fs::read_to_string(path).ok()?;
+    serde_json::from_str(&content).ok()
+}
+
+fn model_config_is_kimi(config: &Value) -> bool {
+    let model_type = config.get("model_type").and_then(Value::as_str);
+    matches!(model_type, Some("kimi_k2") | Some("kimi_k25"))
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/tokenizer/src/kimi_k2_tokenizer.rs:80]: there is no kimi_k26. See https://huggingface.co/moonshotai/Kimi-K2.6/blob/main/config.json#L15

File: crates/tokenizer/src/kimi_k2_tokenizer.rs [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+fn value_mentions_kimi_tokenizer(value: &Value) -> bool {
+    match value {
+        Value::String(s) => s.contains("tokenization_kimi"),
+        Value::Array(values) => values.iter().any(value_mentions_kimi_tokenizer),
+        Value::Object(map) => map.values().any(value_mentions_kimi_tokenizer),
+        _ => false,
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/tokenizer/src/kimi_k2_tokenizer.rs:88]: Fixed in 9c488299 —  now uses dotted-segment match via `split('.').any(...)`. Substring false-positives (`tokenization_kimi_v2` etc.) are covered by a new regression test.

File: crates/tokenizer/src/kimi_k2_tokenizer.rs [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    added_tokens: &mut HashMap<String, TokenIdType>,
+    base_vocab_size: usize,
+) {
+    let Ok(start) = TokenIdType::try_from(base_vocab_size) else {
+        return;
+    };
+
+    let occupied_ids: HashSet<TokenIdType> = added_tokens.values().copied().collect();
+    for offset in 0..NUM_RESERVED_SPECIAL_TOKENS {
+        let id = start + offset as TokenIdType;
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/tokenizer/src/kimi_k2_tokenizer.rs:62]: Skipping. Overflow requires `start > u32::MAX - 255`, i.e. BPE vocab > 4.29B entries — physically unreachable (Kimi's is ~160k).

File: crates/tokenizer/src/tiktoken.rs [hunk 4]
```diff
@@ -213,33 +219,51 @@ impl TiktokenTokenizer {
...
-        // 2. Parse tokenizer_config.json from the same directory
         let dir = tiktoken_path
             .parent()
             .ok_or_else(|| Error::msg("Cannot determine parent directory of tiktoken file"))?;
-        let config_path = dir.join("tokenizer_config.json");
-        let config = if config_path.exists() {
-            load_tiktoken_config(&config_path)?
+        let mut config = load_tiktoken_config_from_dir(dir)?;
+
+        let pattern = if kimi_k2_tokenizer::matches_dir(dir) {
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/tokenizer/src/tiktoken.rs]: Fixed in 9c488299. `load_tiktoken_config_from_dir` now returns `(TiktokenConfig, Option<Value>)` and `kimi_k2_tokenizer::matches` takes the pre-parsed value, so `tokenizer_config.json` is read once. `config.json` fallback still does its own read (different file).

File: crates/tokenizer/src/tiktoken.rs [hunk 5]
```diff
@@ -62,6 +59,15 @@ fn load_tiktoken_config(config_path: &Path) -> Result<TiktokenConfig> {
...
 }
 
+fn load_tiktoken_config_from_dir(dir: &Path) -> Result<TiktokenConfig> {
+    let config_path = dir.join("tokenizer_config.json");
+    if config_path.exists() {
+        load_tiktoken_config(&config_path)
+    } else {
+        Ok(TiktokenConfig::default())
+    }
+}
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/tokenizer/src/tiktoken.rs]: Added in 70002193 — `test_load_tiktoken_config_from_dir_missing_file` covers the new `(default, None)` branch.

File: crates/tokenizer/src/tiktoken.rs [hunk 6]
```diff
@@ -491,6 +508,17 @@ impl Decoder for TiktokenTokenizer {
...
+/// Detect tiktoken's "unknown token id" error so we can surface a clean error
+/// instead of letting the lossy-decode fallback panic on a missing key.
+///
+/// We match on the `Display` string because tiktoken-rs's `DecodeKeyError` lives
+/// in a private `vendor_tiktoken` module and isn't re-exported (as of 0.9.1),
+/// so a typed `downcast_ref` is not available. The message format is stable —
+/// see `vendor_tiktoken::DecodeKeyError::fmt` upstream.
+fn is_unknown_tiktoken_decode_error(err: &Error) -> bool {
+    err.to_string().starts_with("Invalid token for decoding:")
+}
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/tokenizer/src/tiktoken.rs:520]: Skipping. The doc comment already names `vendor_tiktoken::DecodeKeyError::fmt` as the source and notes it isn't re-exported. A regression test would assert against the same private upstream string, so it would break in lockstep with the guard rather than detect drift earlier — no signal gained.

**Comment by key4ng** [comment]
## BFCL non-live — pure vLLM vs SMG (pre-PR) vs SMG (this PR)

Ran the BFCL non-live suite (1390 cases) against Kimi-K2.5 (TP=8 on 8× H200, vLLM v0.21.0-cu129) in three configurations:

| | Configuration | SMG SHA | Non-Live Overall |
|---|---|---|---:|
| **S1** | pure vLLM OpenAI server (no SMG) | — | 87.79% |
| **S2** | SMG fronting vLLM gRPC, pre-PR baseline | `5d154b67` (main 5/4) | 87.92% |
| **S4** | SMG fronting vLLM gRPC, **this PR** | `70002193` | **87.88%** |

S2 ↔ S4 delta: **−0.04 pp / −1 correct out of 1390**. Within noise — a same-build same-suite rerun on a side branch saw 0.16 pp drift, so anything under ±0.5 pp is the floor. PR is effectively a no-op on BFCL non-live, as expected: the decode panic and Han-aware pre-tokenization that this PR fixes aren't exercised by FC-mode English-only tool-calling traffic.

### Per-category

| Category | Total | S1 | S2 | S4 | S4 − S2 |
|---|---:|---:|---:|---:|---:|
| simple_python      | 400 | 380 (95.00%) | 380 (95.00%) | 378 (94.50%) | −2 |
| simple_java        | 100 |  63 (63.00%) |  62 (62.00%) |  62 (62.00%) |  0 |
| simple_javascript  |  50 |  30 (60.00%) |  32 (64.00%) |  32 (64.00%) |  0 |
| multiple           | 200 | 189 (94.50%) | 189 (94.50%) | 187 (93.50%) | −2 |
| parallel           | 200 | 189 (94.50%) | 187 (93.50%) | 186 (93.00%) | −1 |
| parallel_multiple  | 200 | 179 (89.50%) | 180 (90.00%) | 183 (91.50%) | +3 |
| irrelevance        | 240 | 206 (85.83%) | 208 (86.67%) | 208 (86.67%) |  0 |

Setup: `bfcl-eval==2026.3.23`, `--max-model-len 32768`, 32 concurrent threads, temperature 0.001 (BFCL default for `Kimi-K2.5-FC`).

<details>
<summary>Commands used</summary>

**S1 — pure vLLM OpenAI server**

```bash
docker run -tid --gpus all --shm-size 10g \
  -v /raid/models:/models --name kimi-k25-vllm-s1 --network host \
  docker.io/vllm/vllm-openai:v0.21.0-cu129-ubuntu2404 \
  /models/Kimi-K2.5 --trust-remote-code --tensor-parallel-size 8 \
  --tool-call-parser kimi_k2 --enable-auto-tool-choice \
  --served-model-name moonshotai/Kimi-K2.5 \
  --gpu-memory-utilization 0.9 --max-model-len 32768 \
  --port 30000 --host 0.0.0.0
```

**S2 / S4 — SMG fronting vLLM gRPC**

```bash
# vLLM gRPC backend (same args for both scenarios; smg-grpc-servicer mounted from the SMG checkout)
docker run -tid --gpus all --shm-size 10g \
  -v /raid/models:/models -v <smg-checkout>/grpc_servicer:/grpc_servicer \
  --name kimi-k25-vllm-grpc --network host --ulimit nofile=65535:65535 \
  --entrypoint /bin/bash docker.io/vllm/vllm-openai:v0.21.0-cu129-ubuntu2404 \
  -c 'pip install -q smg-grpc-proto grpcio-health-checking grpcio-reflection &&
      cd /grpc_servicer && pip install -q --no-deps -e . && cd / &&
      exec python3 -m vllm.entrypoints.grpc_server \
        --model /models/Kimi-K2.5 --trust-remote-code --tensor-parallel-size 8 \
        --served-model-name moonshotai/Kimi-K2.5 --gpu-memory-utilization 0.9 \
        --max-model-len 32768 --host 0.0.0.0 --port 50051'

# SMG router (built from the target SHA via `maturin develop --release` in bindings/python)
python -m smg.launch_router --host 0.0.0.0 --port 30000 \
  --worker-urls grpc://127.0.0.1:50051 --model-path /raid/models/Kimi-K2.5 \
  --tool-call-parser kimik2 --reasoning-parser kimi_k25 \
  --history-backend memory --policy round_robin --log-level info
```

**BFCL generate + evaluate (same for every scenario; only `OPENAI_BASE_URL` and `--result-dir` change)**

```bash
OPENAI_BASE_URL=http://localhost:30000/v1 OPENAI_API_KEY=EMPTY \
  bfcl generate --model moonshotai/Kimi-K2.5-FC --test-category non_live \
    --backend vllm --skip-server-setup --num-threads 32 \
    --result-dir s4/result --allow-overwrite

OPENAI_BASE_URL=http://localhost:30000/v1 OPENAI_API_KEY=EMPTY \
  bfcl evaluate --model moonshotai/Kimi-K2.5-FC --test-category non_live \
    --result-dir s4/result --score-dir s4/score
```

</details>

**Comment by key4ng** [comment]
## E2E verification of the panic fix

The PR's stated bug: "Decoding a sequence containing an unnamed reserved-token ID falls through to `_decode_native_and_split`, which indexes `special_tokens_decoder[&token]` directly — **panic** on the missing key."

I reproduced this end-to-end through SMG's `/v1/detokenize` endpoint with the same vLLM gRPC backend, varying only which SMG binary is in front of it.

### Identical payload (both scenarios)

```http
POST /v1/detokenize
Content-Type: application/json

{"model":"moonshotai/Kimi-K2.5","tokens":[163700],"skip_special_tokens":false}
```

`163700` is an unnamed slot in Kimi's 256-token reserved range (163585–163840) that `tokenization_kimi.py` synthesizes but `tokenizer_config.json` doesn't enumerate.

### Pre-PR SMG (`5d154b67`)

HTTP connection drops; SMG log shows the exact panic the PR description names:

```
thread 'tokio-rt-worker' (220226) panicked at
/home/ubuntu/.cargo/registry/src/index.crates.io-.../tiktoken-rs-0.9.1/src/patched_tiktoken.rs:82:64:
no entry found for key
```

### With this PR (`70002193`)

```json
{"text":"<|reserved_token_163700|>"}
```

Decodes to the synthetic name `apply_reserved_special_tokens` injects. No panic. Sequences of multiple unnamed reserved IDs also decode cleanly.

### Other code paths the PR touches

Same SMG before/after, hitting `/v1/tokenize` and `/v1/detokenize` with neighboring payloads:

| Probe | Pre-PR | This PR |
|---|---|---|
| Decode `163588` (named `<\|im_assistant\|>`) | `<\|im_assistant\|>` | `<\|im_assistant\|>` ✅ (no regression on named slots) |
| Tokenize `"你好world hello there"` | `[33845, 23617, 40493, 1459]` | `[33845, 23617, 40493, 1459]` |
| Tokenize `"hello's world"` | `[22931, 1048, 2695]` | `[22931, 1048, 2695]` |

Han and contraction tokenization happen to land on the same token IDs for these short inputs — the regex swap matters in edge cases the PR's unit tests cover more rigorously.

### Unit tests added by this PR

```
$ cd <pr-head> && cargo test -p llm-tokenizer --lib kimi_k2_tokenizer
running 6 tests
test kimi_k2_tokenizer::tests::substring_does_not_falsely_match_kimi   ... ok
test kimi_k2_tokenizer::tests::matches_via_model_type_kimi_k25         ... ok
test kimi_k2_tokenizer::tests::matches_via_model_type_kimi_k2          ... ok
test kimi_k2_tokenizer::tests::uses_kimi_pattern_for_contractions      ... ok
test kimi_k2_tokenizer::tests::reserved_special_tokens_are_synthesized ... ok
test kimi_k2_tokenizer::tests::han_input_round_trips_through_kimi_pattern ... ok

test result: ok. 6 passed; 0 failed; 0 ignored; 0 measured; 128 filtered out
```

On PR base (`76d9c180`), the same filter runs **0 tests** — the `kimi_k2_tokenizer` module is new in this PR.
